Update command injection docs (#741)
returntocorp/semgrep-docs: docs/cheat-sheets/python-code-injection.mdx

# Code injection prevention for Python

<CodeInjectionIntro />

## 1. Executing/evaluating code

### 1.A. Executing code with **exec**

`exec()` function supports dynamic execution of Python code.
`exec()` can be dangerous if used to execute dynamic content.
If this content can be input from outside the program, this may be a code injection vulnerability.

Example:

```python
# Value supplied by user
user_input = "');import requests;requests.get('localhost:3000');print('"

# Vulnerable
exec("foobar('{}')".format(user_input))
```

#### References

- [**exec** documentation](https://docs.python.org/3/library/functions.html#exec)

#### Mitigation

Do not use `exec()` for non literal values. Alternatively:

- Ensure executed content is not definable by external sources.
- If it’s not possible, strip everything except alphanumeric characters from an input provided for the command string and arguments.
- Don’t try to make `exec` safe with tricks like `{'__builtins__':{}}`

#### Semgrep rule

<LinkToRegistryRule ruleId="python.lang.security.audit.exec-detected.exec-detected" />

### 1.B. Evaluating code with **eval**

> `eval()` function evaluates string value as a Python expression.
> `eval()` can be dangerous if used to evaluate dynamic content.
> If this content can be input from outside the program, this may be a code injection vulnerability.

Author: ```suggestion
`eval()` function supports dynamic execution of Python code.
`eval()` can be dangerous if it is used to execute dynamic content (non literal content).
If this dynamic content has an input controllable by a user, it can cause a code injection vulnerability.
```

Example:

```python
# Value supplied by user
user_input = "__import__('code').InteractiveInterpreter().runsource('import requests;requests.get(\'localhost:3000\')')"

# Vulnerable
eval(user_input)
```

#### References

- [**eval()** documentation](https://docs.python.org/3/library/functions.html#eval)

#### Mitigation

> Do not use `eval()`. Alternatively:
> 
> - Ensure evaluated content is not definable by external sources.
> - If it’s not possible, strip everything except alphanumeric characters from an input provided for the command string and arguments.
> - Don’t try to make `eval` safe with tricks like `{'__builtins__':{}}`

Author: ```suggestion
The actions below can help you to mitigate this potential vulnerability:

- Don't use `eval()` for non literal values.
- If you need to use `eval()` with non literal values, ensure that executed content is not definable by external sources.
- If the previous action is not possible, strip everything except alphanumeric characters from an input provided for the command string and arguments.
- Don’t try to make `eval` safe with tricks such as `{'__builtins__':{}}`.
```

#### Semgrep rule

<LinkToRegistryRule ruleId="python.lang.security.audit.eval-detected.eval-detected" />

### 1.C. Accepting logging configuration with **logging.config.listen()**

`logging.config.listen()` starts up a socket server on the specified port, and listens for new configurations.

> Because portions of the logging configuration are passed through `eval()`, use of this function may open its users to a security risk.

Author: ```suggestion
As the `logging.config.listen()` configuration is always passed through `eval()`, the use of this function can lead to a security risk.
```

> While the function only binds to a socket on localhost, and so does not accept connections from remote machines, there are scenarios where untrusted code could be run under the account of the process which calls `listen()`.

Author: ```suggestion
While the function only binds to a socket on localhost, and so does not accept connections from remote machines, there are scenarios where untrusted code can potentially run under the account of the process which calls `listen()`.
```

Example:

```python
# Server example: starting up a socket server on 9999 port, and listening for new configurations.
import logging
import logging.config

logging.config.fileConfig('logging.conf')
t = logging.config.listen(9999)
t.start()

# Client example: sending configuration from `data_to_send` variable to localhost:9999
import socket, sys, struct

> # Config example: print("pwned") will be evaluated and "pwned" will be printed to the console

Author: ```suggestion
# Config example: print("pwned") is evaluated and "pwned" is printed to the console
```

data_to_send = """
[loggers]
keys=root

[handlers]
keys=hand01

[formatters]
keys=form01

[logger_root]
level=NOTSET
handlers=hand01

[handler_hand01]
class=StreamHandler
level=NOTSET
formatter=form01
args=(print("pwned"),)

[formatter_form01]
format=F1 %(asctime)s %(levelname)s %(message)s
datefmt=
class=logging.Formatter
"""

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(('localhost', 9999))
s.send(struct.pack('>L', len(data_to_send)))
s.send(data_to_send)
s.close()
```

#### References

- [**logging.config.listen()** documentation](https://docs.python.org/3/library/logging.config.html#logging.config.listen)

#### Mitigation

> Verify what is sent across the socket. Alternatively:
> 
> - To avoid the risk, use the verify argument to `logging.config.listen()` to prevent unrecognized configurations from being applied. This could be done by encrypting and/or signing what is sent across the socket, such that the verify callable can perform signature verification and/or decryption.

Author: ```suggestion
- Verify what is sent across the socket.
- Alternatively: To avoid the risk, use the verify argument to `logging.config.listen()` to prevent applying unrecognized configurations. This can be done by encrypting or signing what is sent across the socket, such that the verify callable can perform signature verification or decryption.
```

### 1.F. Running **subinterpreter** from regression tests package

`run_in_subinterp` is a function from regression tests package for Python that runs code in subinterpreter.
This is dangerous if external data can reach this function call because it allows a malicious actor to run arbitrary Python code.

Example:

```python
import _testcapi

# Value supplied by user
user_input = "print('pwned')"

# Vulnerable
_testcapi.run_in_subinterp(user_input)

from test import support

# Value supplied by user
user_input = "print('pwned')"

# Vulnerable
support.run_in_subinterp(user_input)
```

#### References

- [**test** module documentation](https://docs.python.org/3/library/test.html)

#### Mitigation

Do not let user input in `run_in_subinterp` function. Alternatively:

- Ensure evaluated content is not definable by external sources.
- If it’s not possible, strip everything except alphanumeric characters from an input provided for the command string and arguments.

#### Semgrep rule

<LinkToRegistryRule ruleId="python.lang.security.audit.dangerous-testcapi-run-in-subinterp" />

## 2. Abusing built-in functions

> ### 2.A. Accessing dictionary with current global/local symbol table

Author: ```suggestion
### 2.A. Accessing dictionary with current global or local symbol table
```

> `globals()` and `locals()` return a dictionary representing the current global/local symbol table. Using non-static data to retrieve values from this table is extremely dangerous because it may allow an attacker to execute arbitrary code on the system. 

Author: ```suggestion
`globals()` and `locals()` return a dictionary representing the current global or local symbol table. Using non-static data to retrieve values from this table is extremely dangerous because it can allow an attacker to execute arbitrary code on the system. 
```

Example:

```python
# Name of the arbitrary function supplied by user
user_input = "Name of the function" 

# Vulnerable call of arbitrary function
function = locals().get(user_input)
function()

# Name of the arbitrary function supplied by user
user_input = "Name of the function"

# Vulnerable call of arbitrary function
function = test1.__globals__[user_input]
function()
```

#### References

- [Vulnerable code example](https://github.com/mpirnat/lets-be-bad-guys/blob/d92768fb3ade32956abd53bd6bb06e19d634a084/badguys/vulnerable/views.py#L181-L186)
- [**locals()** documentation](https://docs.python.org/3/library/functions.html#locals)
- [**globals()** documentation](https://docs.python.org/3/library/functions.html#globals)

#### Mitigation

> Do not access global/local symbol tables. Refactor your code not to use `globals()` and `locals()`.

Author: ```suggestion
Do not access global or local symbol tables. Refactor your code not to use `globals()` and `locals()`.
```

#### Semgrep rule

<LinkToRegistryRule ruleId="python.lang.security.dangerous-globals-use.dangerous-globals-use" />

### 2.B. Dynamically updating and accessing code annotations

Annotations passed to `typing.get_type_hints()` are evaluated in `globals` and `locals` namespaces.

> Make sure that no arbitrary value can be written as the annotation and passed to the `typing.get_type_hints` function.

Author: ```suggestion
Ensure that no arbitrary value can be written as the annotation and passed to the `typing.get_type_hints` function.
```

Example:

```python
from typing import get_type_hints

class C:
    member: int = 0

def smth():
    # Changing annotation for `member` property of class C
    C.__annotations__["member"] = "print('pwn')"

    # Annotations are evaluated and `print('pwn')` code gets executed
    get_type_hints(C)
```

#### References

- [**typing.get_type_hints** documentation](https://docs.python.org/3/library/typing.html#typing.get_type_hints)

#### Mitigation

Do not programmatically rewrite code annotations. Alternatively:

> - Ensure that annotations are not definable by external sources.

Author: ```suggestion
- Ensure that annotations are not controllable by external sources.
```
